File: src/test/scripts/suite/semantic/function/name/declaration.js
/*
 * Copyright (c) André Bargull
 * Alle Rechte vorbehalten / All Rights Reserved.  Use is subject to license terms.
 *
 * <https://github.com/anba/es6draft>
 */
const {
  assertSame, assertTrue, assertDataProperty
} = Assert;

function assertFunctionName(f, name) {
  return assertDataProperty(f, "name", {value: name, writable: false, enumerable: false, configurable: true});
}

function assertClassName(c, name) {
  assertTrue(c.hasOwnProperty("name"));
  return assertSame(name, c.name);
}

// function/generator/class declaration
(function() {
  function f() {}
  assertFunctionName(f, "f");

  function* g() {}
  assertFunctionName(g, "g");

  class ClassWithoutName {}
  assertFunctionName(ClassWithoutName, "ClassWithoutName");

  class ClassWithName { static get name() { return "OwnName" } }
  assertClassName(ClassWithName, "OwnName");

  class ClassWithComputedName { static get ["name"]() { return "OwnName" } }
  assertClassName(ClassWithComputedName, "OwnName");
})();

File: src/main/java/com/github/anba/es6draft/ast/Declaration.java
/**
 * Copyright (c) André Bargull
 * Alle Rechte vorbehalten / All Rights Reserved.  Use is subject to license terms.
 *
 * <https://github.com/anba/es6draft>
 */
package com.github.anba.es6draft.ast;

/**
 * <h1>13 ECMAScript Language: Statements and Declarations</h1>
 * <ul>
 * <li>13.2 Block
 * </ul>
 */
public abstract class Declaration extends StatementListItem {
    protected Declaration(long beginPosition, long endPosition) {
        super(beginPosition, endPosition);
    }

    /**
     * Returns {@code true} if this declaration creates an immutable binding.
     * 
     * @return {@code true} if a <code>const</code> declaration
     */
    public abstract boolean isConstDeclaration();
}
